### PR TITLE
Add `<ssr-keep>` + better hydration support to `@bolt/element`

### PR DESCRIPTION
**Pull request by sghoweri**
## Jira
http://vjira2:8080/browse/BDS-2106

## Summary
Ports over the existing `<ssr-keep>`-related hydration logic from `@bolt/core-v3.x` into `@bolt/element`.

## Details
This update was required after running into timing / race condition-related rendering issues once most (all) UI in Bolt started to get loaded and rendered asynchronously.

## How to test
Review code (gut check). Anything else?

All the work actually using this in BoltElement-based components has been split into it's own separate PR so you might need to reference https://github.com/boltdesignsystem/bolt/pull/1839 if you want to see this in action...

### CODE REVIEW

File: packages/testing/testing-helpers/index.js [hunk 1]
```diff
@@ -28,6 +29,45 @@ export const vrtDefaultConfig = {
...
 };
 
+/**
+ * Takes the web component markup passed in, waits for it to render, then returns back the rendered inner and outer HTML
+ * @param {*} componentTag - the web component HTML tag being rendered
+ * @param {String} html - the actual HTML being rendered
+ * @param {*} page - puppeteer page instance to render within
+ * @param {Array} extraSelectors - extra web component selectors to wait to finish rendering before returning
+ */
+export async function renderWC(componentTag, html, page) {
```
Comment (danielamorse) [inline packages/testing/testing-helpers/index.js:39]: Brilliant. I had begun working on something like a while back but this is a much nicer implementation. Goodbye to all that boilerplate Jest code! I'll include it in the next round of Component Generator updates.

File: packages/base-element/src/BoltElement.js [hunk 2]
```diff
@@ -13,12 +13,56 @@ import {
 @conditionalShadowDom()
 @jsonSchemaProps()
 class BoltElement extends Slotify {
+  async connectedCallback() {
+    // wait for Bolt's SSR hydration-related helper elements to be ready before continuing
+    await customElements.whenDefined('ssr-keep');
+
+    // for some very strange reason, waiting for `replace-with-grandchildren` AND `replace-with-children` fails but waiting for just one or the other works
```
Comment (danielamorse) [inline packages/base-element/src/BoltElement.js:20]: Fails how?

File: packages/testing/testing-helpers/index.js (hunk 1 above)
Comment (sghoweri) [reply packages/testing/testing-helpers/index.js:39]: Thanks! 👏

File: packages/base-element/src/BoltElement.js (hunk 2 above)
Comment (sghoweri) [reply packages/base-element/src/BoltElement.js:20]: It waits for that promise to resolve indefinitely (ie. never resolves and the Jest tests ends up timing out).

My suspicion is that it's due to `<replace-with-grandchildren>` extending `<replace-with-children>` and we miss the extra promise... but don't quote me on that. 🤷‍♂️

**Comment by sghoweri** [comment]
PR was released with v2.23.0

**Pull request merged**